Hashtable code improvements

Explorer09 opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary

This PR introduces four commits to improve Hashtable implementation, focusing on safer size boundaries, cleaner consistency checks, and a more compact lookup table representation.

### Changes

**Commit 1:** Extends the `OEISprimes[]` table to include primes up to (2^64 - 59), enabling support for larger hashtable sizes.

**Commit 2:** Moves size-related assertions into `Hashtable_isConsistent()`, tightens grow/shrink conditions to support hashtable sizes up to `SIZE_MAX / 7` without arithmetic overflow, introduces an overflow guard for cases where `sizeof(HashtableItem) < 7`, and makes `Hashtable_setSize()` private by changing it from public API to `static`.

**Commit 3:** Replaces the static `OEISprimes` lookup table with a compact `primeDiffs`-based calculation using OEIS A013603 (differences from 2^n to nearest prime ≤ 2^n), reducing memory footprint from ~512 bytes to ~64 bytes for a runtime-computed lookup model.

Additional modifications include:
- Load-factor growth condition changed from `items >= size / 2` to `items >= size * 7 / 10`
- Growth strategy changed from doubling to incrementing by 1
- Shrink trigger changed from `8 * items < size` to `items < size / 8`
- Shrink target changed from `size / 3` to `size * 3 / 8`

### Commit Structure

The PR contains a structural issue noted in review: **Commit 3 (lookup table shrinking) should be split into a separate PR**. The first three commits are logically independent and can be cherry-picked without the fourth; Commit 3 depends on Commits 1–2, but Commits 1–2 do not depend on Commit 3. The reviewer requested this separation to decouple the size-safety improvements from the optimization debate about precomputed vs. runtime-computed lookup tables.

### Implementation Assessment

The solution is clean with appropriate invariant checks. However, there is unresolved discussion about the lookup table approach: trading a larger precomputed static table (~512 bytes) for simpler/faster runtime access (precomputation via macro) versus maintaining a smaller runtime-computed table (~64 bytes) to reduce binary footprint. The author prioritizes htop's compactness, but has offered to exclude the final commit if maintainers prefer precomputation, suggesting flexibility on the final decision.

### Recommended Action

Merge Commits 1–2 as-is. Separate Commit 3 into its own PR to allow independent evaluation of the lookup table optimization strategy.

<!-- review_stack_entry_start -->

[![Review Change Stack](https://storage.googleapis.com/coderabbit_public_assets/review-stack-in-coderabbit-ui.svg)](https://app.coderabbit.ai/change-stack/htop-dev/htop/pull/1842?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

<!-- review_stack_entry_end -->

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

BenBE commented on Hashtable.c:
```diff
@@ -221,15 +226,12 @@ void Hashtable_put(Hashtable* this, ht_key_t key, void* value) {
-      if (SIZE_MAX / 2 < this->size)
-         CRT_fatalError("Hashtable: size overflow");
-
+   if (10 * this->items > 7 * this->size)
```
TIA.

BenBE commented on Hashtable.c:
```diff
@@ -224,11 +224,12 @@ void Hashtable_put(Hashtable* this, ht_key_t key, void* value) {
+      if (sizeof(HashtableItem) < 10 && SIZE_MAX / 10 < this->size)
          CRT_fatalError("Hashtable: size overflow");
 
       Hashtable_setSize(this, 2 * this->size);
```
On first glance this should work. Not tested in detail though. But given that `setSize` rounds up to the next prime anyway this should likely work.

BenBE commented on Hashtable.c:
```diff
@@ -221,21 +222,21 @@ void Hashtable_put(Hashtable* this, ht_key_t key, void* value) {
-      Hashtable_setSize(this, 2 * this->size);
-   }
+   if (this->items >= this->size * 7 / 10)
+      Hashtable_setSize(this, this->size + 1);
```
Yes, but once the items get near the size because it can't allocate any more buffer space, you could at some point reach items == size, and thus the next insert will fail due to no more space allocated.

Instead when nearing the maximum capacity we should fall of to a more linear allocation regime …

Explorer09 replied:
@BenBE When it "can't allocate any more buffer space" htop will exit, because of the `xCalloc` call.

The case where items == size can only happen on small sizes such as 2 or 3 (the minimum size is 7 now, so the sizes of 2 and 3 are theoretical situations), but even when that happens, the next `Hashtable_put` call will always grow the buffer. Thus there's no problem here.

BenBE replied:
I was more thinking for very large allocations. Will have to take a closer look after New Year's …

BenBE replied:
I think we arrived at this issue before … IIRC.

Maybe inhibit shrinking while we try to insert items …

BenBE commented:
Multiple reasons:
- Reducing runtime code complexity (basically a follow-up to PR #1909)
- No necessity to safe every last byte; in particular with static (constant) program memory.
- Single-byte access is somewhat less efficient here (additional movzx compared to plain memory read)

Explorer09 committed:
Hashtable: Extend OEISprimes[] to up to (2^64 - 59)

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Improve Hashtable buffer grow and shrink conditions

* Move assertions about hash table sizes to Hashtable_isConsistent() so
  they can be checked in all Hashtable methods.
* Slightly improve conditionals of growing and shrinking the "buckets"
  buffer. Specifically the calculations are now less prone to
  arithmetic overflow and can work with Hashtable.size value up to
  (SIZE_MAX / 7). (Original limit was (SIZE_MAX / 10)).
* If `Hashtable.size > SIZE_MAX / sizeof(HashtableItem)`, allow the
  compiler to optimize out one conditional of checking overflow.
  (The buffer allocation would still fail at xCalloc() in that case.)
* Hashtable_setSize() is now a private method.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 committed:
Hashtable: Shrink nextPrime() lookup table size

The lookup table now codes the difference between 2^n to the nearest
prime not greater than 2^n (i.e. https://oeis.org/A013603 ).

With the change of the lookup table, (2^64 - 59) has been removed. It is
believed that such removal won't cause practical problems as the number
is very close to SIZE_MAX and a system is unlikely to succeed in
allocating a memory block _that_ huge.

Signed-off-by: Kang-Che Sung <explorer09@gmail.com>

Explorer09 commented:
> Multiple reasons:
> 
> * Reducing runtime code complexity (basically a follow-up to PR [Allow for optimizing out bound check on 64 bit systems #1909](https://github.com/htop-dev/htop/pull/1909))
> * No necessity to safe every last byte; in particular with static (constant) program memory.
> * Single-byte access is somewhat less efficient here (additional movzx compared to plain memory read)

I don't like the reasoning, although I think such debate is going to be like [arguing the color of a bikeshed](https://bikeshed.com/), and thus I'm not going to change the decisions of the maintainers. You have the choice. I've rebased this PR so you can leave out the lookup table shrink.

Anyway, here's my motivation: My vision of htop is that it should be a compact tool for process monitoring even though it comes with a fancy, terminal UI. That means if there's any chance to reduce memory footprint for the htop program itself, I would wish it could be done (unless there's a larger performance tradeoff with reducing the code size). The memory is better left for large, server applications that need them. In my opinion, it's better for htop to be slightly slower (i.e. take slightly more CPU time) for certain tasks if doing them faster would consume memory that could be essential for server apps.
